python/cuda_parallel/cuda/parallel/experimental/_cccl.py

```diff
@@ -0,0 +1,209 @@
+class TypeInfo(ctypes.Structure):
+    _fields_ = [
+        ("size", ctypes.c_int),
+        ("alignment", ctypes.c_int),
+        ("type", TypeEnum),
+    ]
+
+
+# MUST match `cccl_op_t` in c/include/cccl/c/types.h
+class Op(ctypes.Structure):
+    _fields_ = [
+        ("type", OpKind),
+        ("name", ctypes.c_char_p),
+        ("ltoir", ctypes.c_char_p),
+        ("ltoir_size", ctypes.c_int),
+        ("size", ctypes.c_int),
+        ("alignment", ctypes.c_int),
+        ("state", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_iterator_t` in c/include/cccl/c/types.h
+class Iterator(ctypes.Structure):
+    _fields_ = [
+        ("size", ctypes.c_int),
+        ("alignment", ctypes.c_int),
+        ("type", IteratorKind),
+        ("advance", Op),
+        ("dereference", Op),
+        ("value_type", TypeInfo),
+        ("state", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_device_reduce_build_result_t` in c/include/cccl/c/reduce.h
+class DeviceReduceBuildResult(ctypes.Structure):
+    _fields_ = [
+        ("cc", ctypes.c_int),
+        ("cubin", ctypes.c_void_p),
+        ("cubin_size", ctypes.c_size_t),
+        ("library", ctypes.c_void_p),
+        ("accumulator_size", ctypes.c_ulonglong),
+        ("single_tile_kernel", ctypes.c_void_p),
+        ("single_tile_second_kernel", ctypes.c_void_p),
+        ("reduction_kernel", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_value_t` in c/include/cccl/c/types.h
+class Value(ctypes.Structure):
+    _fields_ = [("type", TypeInfo), ("state", ctypes.c_void_p)]
+
+
+def _type_to_enum(numba_type):
+    mapping = {
+        types.int8: TypeEnum.INT8,
+        types.int16: TypeEnum.INT16,
+        types.int32: TypeEnum.INT32,
+        types.int64: TypeEnum.INT64,
+        types.uint8: TypeEnum.UINT8,
+        types.uint16: TypeEnum.UINT16,
+        types.uint32: TypeEnum.UINT32,
+        types.uint64: TypeEnum.UINT64,
+        types.float32: TypeEnum.FLOAT32,
+        types.float64: TypeEnum.FLOAT64,
+    }
+    if numba_type in mapping:
+        return mapping[numba_type]
+    return TypeEnum.STORAGE
+
+
+# TODO: replace with functools.cache once our docs build environment
+# is upgraded to at least Python 3.9
+@functools.lru_cache(maxsize=None)
+def _numba_type_to_info(numba_type):
+    context = cuda.descriptor.cuda_target.target_context
+    value_type = context.get_value_type(numba_type)
+    size = value_type.get_abi_size(context.target_data)
+    alignment = value_type.get_abi_alignment(context.target_data)
+    return TypeInfo(size, alignment, _type_to_enum(numba_type))
+
+
+@functools.lru_cache(maxsize=None)
+def _numpy_type_to_info(numpy_type):
+    numba_type = numba.from_dtype(numpy_type)
+    return _numba_type_to_info(numba_type)
+
+
+def _device_array_to_cccl_iter(array):
+    dtype = array.dtype
+    info = _numpy_type_to_info(dtype)
+    # Note: this is slightly slower, but supports all ndarray-like objects as long as they support CAI
```

Comment:
Now that we have the nice formatting (ruff), I'd move this comment right above the
```
array.__cuda_array_interface__["data"][0],
```
line.